File: Array/Queue/queue.py
"""
Queue is Linear datastructures which allows insertion and deletion of the element only from one end.
It follows First In First Out principle.It has following three types.
1) Circular Queue
2) Priority Queue
3) Double Ended Queue (Dequeue)
It contains REAR and FRONT. The elements are inserted from the REAR and deleted from the FRONT.
It has the following operations associated:
1) Enqueue - Insert element in the queue.
2) Dequeue - Delete element form the queue.
3) Front - Returns the front element without removing.
4) isEmpty - Checks if the queue is empty or not.
5) isFull - Checks if the queue is Full or not.
"""

QUEUE = []
MAX = 10
FRONT = -1
REAR = -1


def is_empty() -> bool:
    """
    STEP 1: START
    STEP 2: IF REAR is -1 THEN GO TO STEP 3 ELSE Go to step 4
    STEP 3: return TRUE
    STEP 4: return FALSE
    STEP 5: END
    """
    if REAR == -1:
        return True
    return False


def is_full() -> bool:
    """
    STEP 1: START
    STEP 2: IF FRONT is MAX THEN Go to STEP 3 ELSE Go to STEP 4
    STEP 3: return TRUE
    STEP 4: return FALSE
    STEP 5: END
    """
    if REAR < MAX:
        return False
    return True


def front():
    """
    STEP 1: START
    STEP 2: IF not is_empty() THEN Go to STEP 3
    STEP 3: return the element at FRONT
    STEP 4: END
    """
    if is_empty():
        return print("Queue is empty.")
    return print(QUEUE[FRONT])


def enqueue():
    """
    STEP 1: START
    STEP 2: Receive the element to be inserted.
    STEP 3: IF not is_full() THEN Go to 4 ELSE Go to 6.
    STEP 4: Increment the REAR Pointer.
    STEP 5: Store the element at REAR Location.
    STEP 6: return overflow message.
    """
    global REAR
    if is_full():
        return print("The Queue is full.")

    value = int(input("Enter the value: "))
    if REAR < MAX:
        REAR += 1
        QUEUE.append(value)
        print("Added to Queue.")


def dequeue():
    global FRONT, REAR
    """
    STEP 1: START
    STEP 2: IF is_empty THEN go to 3 ELSE go to 4
    STEP 3: return Empty message
    STEP 4: print the element stored at FRONT location.
    STEP 5: increment the FRONT pointer
    STEP 6: IF FRONT == REAR THEN go to 7
    STEP 7: SET FRONT AND REAR = -1
    STEP 8: END
    """
    if is_empty():
        return print("Queue is empty!")
    else:
        FRONT += 1
        print(f"\nRemoved {QUEUE[FRONT]}")
    if FRONT > REAR:
        FRONT = -1
        REAR = -1


def display():
    for item in QUEUE:
        print(item)


def main():
    choice = int(input("\n1.Enqueue\n2.Dequeue\n3.Is Full\n4.Is Empty\n5.Peek\n6.Display\nEnter your choice: "))
    if choice == 1:
        enqueue()
    elif choice == 2:
        dequeue()
    elif choice == 3:
        is_full()
    elif choice == 4:
        is_empty()
    elif choice == 5:
        front()
    elif choice == 6:
        display()
    else:
        print("Choose from above options only!")


if __name__ == "__main__":
    while True:
        main()

File: Array/Queue.c
#include <stdio.h>
#include <stdbool.h>
int FRONT = -1;
int REAR = -1;
int queue[10];

/*
    Queue is a linear data structure in which the elements are added from one end and deleted from another end. It follows First In First Out Principle. It has following operations:
    1) Insert (Enqueue)
    2) Delete (Dequeue)
    3) Display
    4) Peek
    5) isEmpty
    6) isFull
*/

void enqueue();
void dequeue();
void display();
void peek();
bool isEmpty();
bool isFull();

int main()
{
    int choice;
    while (1)
    {
        printf("\n1.Enqueue\n2.Dequeue\n3.Display\n4.Peek\n5.Exit\nSelect the above choice:- ");
        scanf("%d", &choice);
        switch (choice)
        {
        case 1:
            enqueue();
            break;
        case 2:
            dequeue();
            break;
        case 3:
            display();
            break;
        case 4:
            peek();
            break;
        case 5:
            return 0;
        default:
            printf("\nPlease choose from given options!");
            break;
        }
    }
}

bool isEmpty()
{
    if (REAR == -1 && FRONT == -1)
        return true;
    else
        return false;
}

bool isFull()
{
    if (REAR == 9)
        return true;
    else
        return false;
}

void enqueue()
{
    if (isFull())
    {
        printf("\nQueue is Full!");
        return;
    }
    printf("\nEnter the element to be enqueued: ");
    if (REAR == -1 && FRONT == -1)
    {
        REAR = 0;
        FRONT = 0;
        scanf("%d", &queue[REAR]);
    }
    else
        scanf("%d", &queue[++REAR]);
}

void dequeue()
{
    if (isEmpty())
    {
        printf("\nQueue is empty!");
        return;
    }
    printf("\nRemoved %d", queue[FRONT++]);

    if (FRONT > REAR)
    {
        FRONT = -1;
        REAR = -1;
    }
}

void peek()
{
    if (isEmpty())
    {
        printf("\nQueue is empty!");
        return;
    }
    printf("\nPeek Element is : %d", queue[FRONT]);
}

void display()
{
    if (isEmpty())
    {
        printf("\nQueue is empty!");
        return;
    }
    for (int i = FRONT; i <= REAR; i++)
    {
        printf("\n%d", queue[i]);
    }
}